rfcs/20210504-kernel-extension-variable-ops.md

```diff
@@ -0,0 +1,218 @@
+// way of performing weights update during the optimizer op. It returns an opaque 
+// LockHolder handle back to plugin. This handle is passed to the Release API for 
+// releasing the locks when the weight update is done.
+TF_CAPI_EXPORT extern void TF_MaybeLockVariableInputMutexesInOrder(
+                                    TF_OpKernelContext* ctx, bool do_lock, bool sparse,
+                                    const int* const input_ids,
+                                    size_t len,
+                                    TF_VariableInputLockHolder** lockHolder,
+                                    TF_Status* status);
+
+// This interface returns out tensor which is updated corresponding to the
+// variable passed with input index.
+TF_CAPI_EXPORT extern void TF_GetInputTensorFromVariable(
+                                        TF_OpKernelContext* ctx, 
+                                        int input,
+                                        bool lock_held,
+                                        bool sparse,
+                                        void (*copyFunc)(
+                                                TF_OpKernelContext * ctx,
+                                                TF_Tensor *source,
+                                                TF_Tensor *dest),
+                                        TF_Tensor** out,
+                                                          TF_Status* status);
+
+// This interface forwards the reference from input to the output tensors
+// corresponding to the indices provided with input_index and output_index
+TF_CAPI_EXPORT extern void TF_OpKernelContext_ForwardRefInputToRefOutput(
+                                                   TF_OpKernelContext* ctx,
+                                                   int32_t input_index,
+                                                   int32_t output_index);
+
+// The API releases the opaque lock handle returned with
+// TF_MaybeLockVariableInputMutexesInOrder API
+TF_CAPI_EXPORT extern void TF_ReleaseVariableInputLockHolder(
+                                  TF_VariableInputLockHolder* lockHolder);
+
+
+```
+
+*Resource Variables*
+ 
+Below mentioned APIs expose functionality in Core TensorFlow to allow plugins
+to do Assign and Update operations to implement different ResourceVariable
+primitives. These are higher level operations which can be easily integrated
+into the plugin by directly calling these APIs from the Compute function of
+these ops in the Plugin. 
+ 
+```
+// Expose higher level Assignment operation for Pluggable vendors to implement
+// in the plugin for Training. The API takes in the context with indices for
+// the input and value tensors. It also accepts the copy functor provided by
+// pluggable vendor to do the copying of the tensors.
+TF_CAPI_EXPORT extern void TF_AssignVariable(TF_OpKernelContext* ctx,
+                               int input_index,
```

Comment:
Hi @jzhoulon , we looked into other ResourceVariableOps. 

1. `ReadVariableOp` is already part of DEVICE_DEFAULT implementation
2. `TensorArrayReadOp` this we haven't looked in detail. From the current look the only Device dependent information is using [LockedRead](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/tensor_array.cc;l=60) where we can memset Tensor to 0.
3. AssignVariable and AssignVariableUpdate are handled  as implementations in the interface.
4. ResourceGather, ResourceGatherNdOp, ResourceScatter and ResourceScatterNdOp follow the pattern of :

`
  void Compute(OpKernelContext* c) override {
    core::RefCountPtr<Var> v;
    OP_REQUIRES_OK(c, LookupResource(c, HandleFromInput(c, 0), &v));
    OP_REQUIRES_OK(c, EnsureSparseVariableAccess<Device, T>(c, v.get()));
    if (is_non_pod_dtype || use_exclusive_lock_) {
      mutex_lock ml(*v->mu());
    << Compute Functor performing the operation >>   
    } else {
      tf_shared_lock ml(*v->mu());
    << Compute Functor performing the operation >>   
    }

`
We implemented these operations where the :
`RefCountPtr<Var>`, processed by LookupResource and EnsureSparseVariableAccess (with locking) is implemented using the TF_GetInputResourceVariable 
and TF_MaybeLockVariableInputMutexesInOrder.

Reply:
Hi @kulinseth . We have went through the resource relevant ops and found that the below ops will use some device specific function, such as concat and split and so on. Not only just the memset. And We roughly put it in below table. Will this RFC consider these ops  like the variable and add some new high level APIs ?


| Resource        | Ops                         |
|-----------------|-----------------------------|
| TensorArray 		    | TensorArrayWrite (V2, V3)   |
|  			   			 		           | TensorArrayRead (V2, V3)    |
|  			   			 		           | TensorArrayPack             |
|  			   			 		           |  TensorArrayGather (V2, V3) |
|  			   			 		           | TensorArrayConcat (V2, V3)  |
|  			   			 		           | TensorArrayUnpack           |
|  			   			 		           | TensorArrayScatter (V2, V3) |
|  			   			 		           | TensorArraySplit (V2, V3)   |
| UnBatchResource | UnbatchKernel               |
|                 | UnbatchGradKernel           |

Reply:
Hi @yanzhang-dev ,  thanks for the list and feedback. We haven't looked at TensorArray ops in detail, but it seems to me we would need more than device specific functionality. We might need to expose the `TF_TensorArray` along with necessary methods like `TF_GetTensorArray`. This might need a separate RFC and is probably outside the scope of what we were looking at here. Similarly for UnbatchResource:

```
// A class encapsulating the state and logic for unbatching tensors.
//
// UnbatchResource keeps two data structures indexed by batch-key: one which has
// the continuations for all concurrent kernels which are waiting for tensors
// and another which has tensors which are waiting for their corresponding
// kernels to run. Whenever a kernel runs, we either grab its tensor if it's
// waiting already, or we insert it in the queue and then look at its tensor to
// see if it can be used to dispatch any stored continuations.
class UnbatchResource : public ResourceBase {
 public:
  explicit UnbatchResource(int32 timeout_micros)
      : timeout_micros_(timeout_micros),
        timeout_enforcer_(new serving::PeriodicFunction(
            [this] { EnforceTimeout(); }, 1000 /* 1 ms */)) {}

  ~UnbatchResource() override {
    // Tear down 'timeout_enforcer_' first, since it accesses other state in
    // this class.
    timeout_enforcer_ = nullptr;
  }
 
```

I agree we can probably expose the `TF_ResourceBase`, `TF_ResourceManager` and others and potentially recreate all this logic on Plugin side.

On a separate note I was curious about these operations, which networks did you encounter these for performance critical sections.?